rust/datafusion/src/execution/physical_plan/hash_aggregate.rs

```diff
@@ -327,120 +278,49 @@ impl RecordBatchReader for GroupedHashAggregateIterator {
-                        col,
-                        accumulators
-                    ),
-                    DataType::UInt8 => update_accumulators!(
-                        array,
-                        UInt8Array,
-                        ScalarValue::UInt8,
-                        col,
-                        accumulators
-                    ),
-                    DataType::UInt16 => update_accumulators!(
-                        array,
-                        UInt16Array,
-                        ScalarValue::UInt16,
-                        col,
-                        accumulators
-                    ),
-                    DataType::UInt32 => update_accumulators!(
-                        array,
-                        UInt32Array,
-                        ScalarValue::UInt32,
-                        col,
-                        accumulators
-                    ),
-                    DataType::UInt64 => update_accumulators!(
-                        array,
-                        UInt64Array,
-                        ScalarValue::UInt64,
-                        col,
-                        accumulators
-                    ),
-                    DataType::Float32 => update_accumulators!(
-                        array,
-                        Float32Array,
-                        ScalarValue::Float32,
-                        col,
-                        accumulators
-                    ),
-                    DataType::Float64 => update_accumulators!(
-                        array,
-                        Float64Array,
-                        ScalarValue::Float64,
-                        col,
-                        accumulators
-                    ),
-                    other => {
-                        return Err(ExecutionError::ExecutionError(format!(
-                            "Unsupported data type {:?} for result of aggregate expression",
-                            other
-                        )).into_arrow_external_error());
+                // for each new key on the map, add an accumulatorSet to the map
+                match map.get(&key) {
+                    None => {
+                        let accumulator_set: AccumulatorSet = self
+                            .aggr_expr
+                            .iter()
+                            .map(|expr| expr.create_accumulator())
+                            .collect();
+                        map.insert(key.clone(), Rc::new(accumulator_set));
                     }
+                    _ => (),
                 };
+
+                // iterate over each non-grouping column in the batch and update the accumulator
+                // for each row
+                for col in 0..aggr_input_values.len() {
+                    let value = get_scalar_value(&aggr_input_values[col], row)
+                        .map_err(ExecutionError::into_arrow_external_error)?;
+
+                    match map.get(&key) {
```

Comment:
It looks like this code is performing a map lookup for the same key once for each aggregate column. It would be more efficient to do the map lookup once per row.

I don't recall whether we have a cargo bench for hash aggregate that we can use to check for performance regressions but maybe we should add one.

Reply:
I agree. I think that this is already the case on [line 345](https://github.com/apache/arrow/pull/7687/files#diff-cd06ba69700f4a62c9176b691402908cL345). 

I was trying to keep this commit semantic free and just focus on simplifying code. However, I can expand the scope and fix these. Let me know what do you think it is best.

Reply:
I'm pretty sure the current code performs one map lookup for row:

```rust
            for row in 0..batch.num_rows() {
                // create grouping key for this row
                create_key(&group_values, row, &mut key)
                    .map_err(ExecutionError::into_arrow_external_error)?;

                if let Some(accumulator_set) = map.get(&key) {
```

However, there are a number of inefficiencies in the current implementation and I was planning on contributing some smaller changes to fix those (such as https://issues.apache.org/jira/browse/ARROW-9550). I like some of the cleanup you have in this PR as well. I'm not sure of the best way to co-ordinate on the changes.

Reply:
you are right, I am sorry for the wrong information. Very well spotted.

1. I can fix this
2. we can scrap this PR
3. you can just go ahead and push your changes and I work on rebasing this on top of your changes

either is fine for me. If you already started your part on the current master, then we close this PR and go ahead with 2 or 3 (just ping me on your PR and I follow-up in this). If not and would prefer to work on top of this code, I can work on this tomorrow to not block you any further. :)

On a non-technical note, I am fine to scrap code I wrote if it is not sufficiently valuable given the circumstances, so no worries on that front whatsoever.

Reply:
I fixed both issues for now.